concrete-core/src/backends/core/private/crypto/ggsw/ciphertext.rs

```diff
@@ -432,4 +462,220 @@ impl<Cont> GgswCiphertext<Cont> {
                 )
             })
     }
+
+    pub fn fill_with_trivial_encryption<Scalar>(&mut self, plaintext: &Plaintext<Scalar>)
+    where
+        Self: AsMutTensor<Element = Scalar>,
+        Scalar: UnsignedTorus,
+    {
+        // We fill the ggsw with trivial glwe encryptions of zero:
+        for mut glwe in self.as_mut_glwe_list().ciphertext_iter_mut() {
+            let mut mask = glwe.get_mut_mask();
+            mask.as_mut_tensor().fill_with_element(Scalar::ZERO);
+        }
+        let base_log = self.decomposition_base_log();
+        for mut matrix in self.level_matrix_iter_mut() {
+            let decomposition = plaintext.0.wrapping_mul(
+                Scalar::ONE
+                    << (<Scalar as Numeric>::BITS
+                        - (base_log.0 * (matrix.decomposition_level().0))),
+            );
+            // We iterate over the rows of the level matrix
+            for (index, row) in matrix.row_iter_mut().enumerate() {
+                let rlwe_ct = row.into_glwe();
+                // We retrieve the row as a polynomial list
+                let mut polynomial_list = rlwe_ct.into_polynomial_list();
+                // We retrieve the polynomial in the diagonal
+                let mut level_polynomial = polynomial_list.get_mut_polynomial(index);
+                // We get the first coefficient
+                let first_coef = level_polynomial.as_mut_tensor().first_mut();
+                // We update the first coefficient
+                *first_coef = first_coef.wrapping_add(decomposition);
+            }
+        }
+    }
+
+    pub fn external_product<C1, C2, Scalar>(
+        &self,
+        output: &mut GlweCiphertext<C1>,
+        glwe: &GlweCiphertext<C2>,
+    ) where
+        Self: AsRefTensor<Element = Complex64>,
+        GlweCiphertext<C1>: AsMutTensor<Element = Scalar>,
+        GlweCiphertext<C2>: AsRefTensor<Element = Scalar>,
+        Scalar: UnsignedTorus,
+    {
+        // We check that the polynomial sizes match
+        ck_dim_eq!(
+            self.poly_size =>
+            glwe.polynomial_size(),
+            output.polynomial_size()
+        );
+        // We check that the glwe sizes match
+        ck_dim_eq!(
+            self.glwe_size() =>
+            glwe.size(),
+            output.size()
+        );
+
+        let mut buffers = FourierBskBuffers::new(self.polynomial_size(), self.glwe_size());
+        let fft_buffers = &mut buffers.fft_buffers;
```

Comment:
I think it would make sense to expose an argument in the method to provide the buffers (you can actually use an `Option` to make it optional). This way, we could use the ones stored in the `CoreEngine`. That would prevent re-allocating them each and every time we call the external product !

Reply:
Right, will do.

Reply:
I'm encountering two issues with this: first, the buffers in CoreEngine seem to be bound to a bootstrapping key, so there are none available for an external product. Second, if the buffers are optional, I still have to allocate some in the body of the function as a default value, just in case, so there is no gain.